File: java/Core/src/com/tapstream/sdk/Event.java
package com.tapstream.sdk;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

import org.json.JSONException;
import org.json.JSONObject;

public class Event {
	private double firstFiredTime = 0;
	private String uid;
	private String name;
	private String encodedName;
	private boolean oneTimeOnly;
	private StringBuilder postData = new StringBuilder();
	private boolean isTransaction = false;
	private String productSku;
	Map<String, Object> customFields = new HashMap<String, Object>();

	public Event(String name, boolean oneTimeOnly) {
		uid = makeUid();
		this.oneTimeOnly = oneTimeOnly;
		setName(name);
	}

	// Only to be used for creating custom purchase events
	public Event(String orderId, String productSku, int quantity) {
		this("", false);
		isTransaction = true;
		this.productSku = productSku;

		addPair("", "purchase-transaction-id", orderId);
		addPair("", "purchase-product-id", productSku);
		addPair("", "purchase-quantity", quantity);
	}

	// Only to be used for creating custom purchase events
	public Event(String orderId, String productSku, int quantity, int priceInCents, String currencyCode) {
		this("", false);
		isTransaction = true;
		this.productSku = productSku; 

		addPair("", "purchase-transaction-id", orderId);
		addPair("", "purchase-product-id", productSku);
		addPair("", "purchase-quantity", quantity);
		addPair("", "purchase-price", priceInCents);
		addPair("", "purchase-currency", currencyCode);
	}

	// Only to be used for creating IAB purchase events
	public Event(JSONObject purchase, JSONObject skuDetails) throws JSONException {
		this("", false);
		isTransaction = true;
		productSku = purchase.getString("productId");
		String orderId = purchase.getString("orderId");

		try {
			String currencyCode = skuDetails.getString("price_currency_code");
			int priceMicros = skuDetails.getInt("price_amount_micros");
			int priceCenti = (int)Math.round(priceMicros / 10000.0);

			addPair("", "purchase-transaction-id", orderId);
			addPair("", "purchase-product-id", productSku);
			addPair("", "purchase-quantity", 1);
			addPair("", "purchase-price", priceCenti);
			addPair("", "purchase-currency", currencyCode);
			
		} catch (JSONException e) {
			// Older versions of the Google Play Store app don't send the currency and amount separately
			addPair("", "purchase-transaction-id", orderId);
			addPair("", "purchase-product-id", productSku);
			addPair("", "purchase-quantity", 1);
		}
	}

	public void addPair(String key, Object value) {
		customFields.put(key, value);
	}

	public String getUid() {
		return uid;
	}

	public String getName() {
		return name;
	}

	public String getEncodedName() {
		return encodedName;
	}

	public boolean isOneTimeOnly() {
		return oneTimeOnly;
	}

	public String getPostData() {
		return postData != null ? postData.toString() : "";
	}

	boolean isPrepared() {
		return firstFiredTime != 0;
	}
	
	void prepare(final Map<String, Object> globalEventParams) {
		// Only record the time of the first fire attempt
		if (firstFiredTime == 0) {
			firstFiredTime = System.currentTimeMillis();
			
			for(Map.Entry<String, Object> entry : globalEventParams.entrySet()) {
				if(!customFields.containsKey(entry.getKey())) {
					customFields.put(entry.getKey(), entry.getValue());
				}
			}

			postData.append(String.format(Locale.US, "&created-ms=%.0f", firstFiredTime));
			
			for(Map.Entry<String, Object> entry : customFields.entrySet()) {
				addPair("custom-", entry.getKey(), entry.getValue());
			}
		}
	}

	boolean isTransaction() {
		return isTransaction;
	}

	void setName(String name) {
		this.name = name.toLowerCase().trim();
		try {
			encodedName = URLEncoder.encode(this.name, "UTF-8").replace("+", "%20");
		} catch (UnsupportedEncodingException e) {
			e.printStackTrace();
		}
	}

	void setNamePrefix(String appName) {
		setName(String.format(Locale.US, "android-%s-purchase-%s", appName, productSku));
	}

	private String makeUid() {
		return String.format(Locale.US, "%d:%f", System.currentTimeMillis(), Math.random());
	}
	
	protected void addPair(String prefix, String key, Object value) {
        String encodedPair = Utils.encodeEventPair(prefix, key, value);
        if(encodedPair != null) {
	        postData.append("&");
	        postData.append(encodedPair);
        }
	}
};

File: cs/Core/Event.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace TapstreamMetrics.Sdk
{
    public sealed class Event
    {
        private static Random rng = new Random();

        private double firstFiredTime = 0;
        private string uid;
        private string name;
        private string encodedName;
        private bool oneTimeOnly;
        private StringBuilder postData = new StringBuilder();
        internal IDictionary<string, object> customFields = new Dictionary<string, object>();

        private bool isTransaction = false;
        private string productId;

        public Event(string name, bool oneTimeOnly)
        {
            uid = MakeUid();
            SetName(name);
            this.oneTimeOnly = oneTimeOnly;
        }

        // This constructor is only to be used for creating custom IAP events.
        public Event(string transactionId, string productId, int quantity, int priceInCents, string currencyCode)
            : this("", false)
        {
            this.productId = productId;
            isTransaction = true;
            AddPair("", "purchase-transaction-id", transactionId);
            AddPair("", "purchase-product-id", productId);
            AddPair("", "purchase-quantity", quantity);
            AddPair("", "purchase-price", priceInCents);
            AddPair("", "purchase-currency", currencyCode);
        }

        // This constructor is only to be used for creating custom IAP events.
        public Event(string transactionId, string productId, int quantity)
            : this("", false)
        {
            this.productId = productId;
            isTransaction = true;
            AddPair("", "purchase-transaction-id", transactionId);
            AddPair("", "purchase-product-id", productId);
            AddPair("", "purchase-quantity", quantity);
        }

        public void AddPair(string key, Object value)
        {
            customFields.Add(key, value);
        }

        public string Uid
        {
            get
            {
                return uid;
            }
        }

        public string Name
        {
            get
            {
                return name;
            }
        }

        public string EncodedName
        {
            get
            {
                return encodedName;
            }
        }

        public bool OneTimeOnly
        {
            get
            {
                return oneTimeOnly;
            }
        }

        public string PostData
        {
            get
            {
                return postData != null ? postData.ToString() : "";
            }
        }

        internal bool IsTransaction
        {
            get
            {
                return isTransaction;
            }
        }

        internal void Prepare(IDictionary<string, object> globalEventParams)
        {
            // Only record the time of the first fire attempt
            if(firstFiredTime == 0)
            {
                TimeSpan t = (DateTime.UtcNow - new DateTime(1970, 1, 1));
                firstFiredTime = t.TotalMilliseconds;

                foreach (string key in globalEventParams.Keys)
                {
                    if (!customFields.ContainsKey(key))
                    {
                        customFields.Add(key, globalEventParams[key]);
                    }
                }

                postData.Append(String.Format("&created-ms={0:0.}", firstFiredTime));

                foreach (string key in customFields.Keys)
                {
                    AddPair("custom-", key, customFields[key]);
                }
            }
        }

        internal void SetName(string eventName)
        {
            name = eventName.ToLower().Trim().Replace(".", "_");
            encodedName = Utils.EncodeString(name);
        }

        internal void SetNamePrefix(string platform, string appName)
        {
            SetName(String.Format("{0}-{1}-purchase-{2}", platform, appName.ToLower().Trim(), productId));
        }

        private string MakeUid()
        {
            return String.Format("{0}:{1}", Environment.TickCount, rng.NextDouble());
        }

        private void AddPair(string prefix, string key, Object value)
        {
            string encodedPair = Utils.EncodeEventPair(prefix, key, value);
            if (encodedPair != null)
            {
                postData.Append("&");
                postData.Append(encodedPair);
            }
        }
    }
}
